### PR TITLE
Code for the llm-finetuning blog

### PR DESCRIPTION
**Pull request by KevinMusgrave**

### CODE REVIEW

File: blog/llm-finetuning/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+```bash
+pip install determined
+```
+
+Then finetune:
+```bash
+det e create distributed.yaml . 
+```
+
+Change configuration options in `distributed.yaml`.
```
Comment (aciborowska) [inline blog/llm-finetuning/README.md]: I'd add a sentence or two about useful parameters to change (e.g., slots_per_trial).

File: blog/llm-finetuning/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+Test your model's generation capabilities:
+
+```bash
+python test_model.py --exp_id <exp_id> --dataset_subset <dataset_subset>
```
Comment (aciborowska) [inline blog/llm-finetuning/README.md:18]: let's make sure to mention that exp_id comes from det e create distributed.yaml.

File: blog/llm-finetuning/distributed.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+name: language modeling distributed
+debug: false
+environment:
+  image: determinedai/environments-dev:python-3.10-pytorch-2.0-deepspeed-0.10.0-smartsim
```
Comment (aciborowska) [inline blog/llm-finetuning/distributed.yaml]: let's change to the default det image. this one is genai specific. We will need to add deepspeed to the requirements.py.

File: blog/llm-finetuning/distributed.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  environment_variables:
+    - NCCL_DEBUG=INFO
+    # - HF_HOME=
```
Comment (aciborowska) [inline blog/llm-finetuning/distributed.yaml]: let's remove it the commented code before merging.

File: blog/llm-finetuning/distributed.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+resources:
+  slots_per_trial: 2
+  resource_pool: A100
```
Comment (aciborowska) [inline blog/llm-finetuning/distributed.yaml]: let's remove the resource pool name. I think in the blog post and/or README.md we should add info about required minimum config. For 1.1B model we should be good with 2xT4s, which are more accessible than A100s.

File: blog/llm-finetuning/distributed.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    batches: 5000
+  metric: eval_accuracy
+  smaller_is_better: false
+hyperparameters:
+  model: "TinyLlama/TinyLlama-1.1B-Chat-v0.4"
+  dataset_subset: "easy"
+  training_args:
+    output_dir: "/tmp/llm_finetuning"
+    max_steps: 5000
+    per_device_train_batch_size: 1
```
Comment (aciborowska) [inline blog/llm-finetuning/distributed.yaml:20]: nit: It would be great if we could increase per_device_train_batch_size, but I'm leaving decision up to you.

File: blog/llm-finetuning/distributed.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    per_device_eval_batch_size: 16
+    fp16: true
+    evaluation_strategy: "steps"
+    eval_steps: 1000
+    logging_strategy: "steps"
+    logging_steps: 1
```
Comment (aciborowska) [inline blog/llm-finetuning/distributed.yaml]: I think this makes the loss very "jumpy" and hard to read. let's default to 100 given 5000 batches to train

File: blog/llm-finetuning/test_model.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import argparse
```
Comment (aciborowska) [inline blog/llm-finetuning/test_model.py:1]: Thanks for adding this script!

File: blog/llm-finetuning/finetune.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    return logits.argmax(dim=-1)
+
+
+def main(training_args, det_callback, hparams):
+    model_name = hparams["model"]
+    model = AutoModelForCausalLM.from_pretrained(model_name)
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    def tokenize(element):
+        formatted = tokenizer.apply_chat_template(
```
Comment (aciborowska) [inline blog/llm-finetuning/finetune.py:31]: I think there is a problem with apply_chat_template. Sorry for noticing this only now, I should have spotted that earlier.

I noticed a warning when running inference, saying that sth like `this tokenizer does not have chat template defined, defaulting to LlamaTokenizer chat template`. When I looked at the [TinyLlama model card](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v0.4), and [their inference script](https://github.com/jzhang38/TinyLlama/blob/main/sft/simple_inference2.py), it looks that they fine tuned using [open-ai chat-ml format](https://github.com/MicrosoftDocs/azure-docs/blob/main/articles/ai-services/openai/includes/chat-markup-language.md), which is different from llama2 format. 

I changed default prompt template to (following [this](https://huggingface.co/docs/transformers/main/chat_templating#advanced-adding-and-editing-chat-templates)):
```
    chat_template = """
{% for message in messages %}
{% if message['role'] == 'user' %}
{{'<|im_start|>user\n' + message['content'].strip() + '<|im_end|>' }}
{% elif message['role'] == 'system' %}
{{'<|im_start|>system\n' + message['content'].strip() + '<|im_end|>' }}
{% elif message['role'] == 'assistant' %}
{{'<|im_start|>assistant\n'  + message['content'] + '<|im_end|>' }}
{% endif %}
{% endfor %}
    """
   tokenizer.chat_template = chat_template
```

and I got a better response from the model now:
```
Correct response:
SELECT date FROM table_name_74 WHERE tournament = "michalovce"


LLM response:
The date of the tournament was held on the 2nd of March 2023.

74. CREATE TABLE table_name_74 (
    date VARCHAR,
    tournament VARCHAR
)

This SQL query creates a new table called `table_name_74` with two columns `date` and `tournament`. The `date` column is a VARCHAR type and the `tournament` column is a VARCHAR type. [...]
```

If I am correct, we will need to re-run fine-tuning jobs. I hope this will also help with `hard` dataset, so it may be worth it to have a better story.

File: blog/llm-finetuning/dataset_utils.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def filter_by_total_length(df, difficulty):
+    if difficulty == "easy":
+        return df[df["total_length"].between(10, 100)]
+    elif difficulty == "medium":
+        return df[df["total_length"].between(101, 200)]
+    elif difficulty == "hard":
+        return df[df["total_length"].between(200, 800)]
+
```
Comment (aciborowska) [inline blog/llm-finetuning/dataset_utils.py]: nit: how about we take up to 10K samples for each dataset (e.g., `df[df["total_length"].between(10, 100)].iloc[:10000]`)?

**Review by aciborowska — APPROVED** [review]
LGTM! Thanks for fixing chat prompt template
